charmcraft/utils.py

```diff
@@ -404,3 +419,107 @@ def get_charm_name_from_path(path: pathlib.Path) -> str:
     if metadata_yaml is None or "name" not in metadata_yaml:
         raise InvalidCharmPathError(path)
     return metadata_yaml["name"]
+
+
+def get_pypi_packages(*requirements: Iterable[str]) -> Set[str]:
+    """Get a set of pypi packages from requirements files.
+
+    :param requirements: An iterable of strings for each requirement.
+    :returns: A set of package names and their requirements (e.g. version numbers)
+    """
+    packages = set()
+    for req in requirements:
+        for line in req:
+            line = line.strip()
+            if line.startswith("-"):
+                continue
+            if match := PACKAGE_LINE_REGEX.match(line):
+                packages.add(match.group(0))
+
+    return packages
+
+
+def get_package_names(packages: Iterable[str]) -> Set[str]:
+    """Get just the names of packages from an iterable of package lines.
+
+    :param packages: An iterable of package lines (e.g. ["abc==1.0.0", "def"])
+    :returns: A set of package names only (e.g. {"abc", "def"})
+    """
+    names = set()
+    for package in packages:
+        if match := PACKAGE_LINE_REGEX.match(package):
+            names.add(match.group(1))
+
+    return names
+
+
+def exclude_packages(requirements: Set[str], *, excluded: Collection[str]) -> Set[str]:
+    """Filter a set of requirements lines by a collection of package names.
+
+    :param requirements: A set of requirements lines (e.g. {"abc==1.0.0", "def>=1.0"})
+    :param excluded: A collection of package names (only) to exclude.
+    :returns A filtered set of requirements, excluding the given packages.
+    """
+    exclusions = set()
+    for requirement in requirements:
+        if match := PACKAGE_LINE_REGEX.match(requirement):
+            if match.group(1) in excluded:
+                exclusions.add(requirement)
+
+    return requirements - exclusions
+
+
+def get_pip_command(
+    prefix: Iterable[str],
+    requirements_files: Collection[pathlib.Path],
+    *,
+    source_deps: Collection[str] = (),
+    binary_deps: Collection[str] = (),
+) -> List[str]:
+    """Build a pip command based on requirements files and dependencies.
+
+    :param prefix: The pip command and any earlier arguments.
+    :param requirements_files: Paths to the requirements files to include.
+    :param source_deps: Additional dependencies that can only be installed from source.
+    :param binary_deps: Dependencies (including from requirements files) allowed for binary install.
+    :returns: A full pip command line
+    """
+    charm_packages = get_pypi_packages(source_deps)
+    binary_packages = get_pypi_packages(binary_deps)
+    requirements_packages = get_pypi_packages(
```

Comment:
Charmcraft only supports pypi packages?  If you have `my-library @ git+https://github.com/...` in a requirements file, it will be silently excluded, right?  

Was it allowed in the old behavior?

Reply:
If a requirements file contains values like that, it'll still be parsed, but it won't be included in consideration of whether the package needs to be included in the `--no-binary` option. Likewise, a requirements file that links directly to a wheel file will still work just as before.